### PR TITLE
Implement Message Counter Synchronization Protocol part

### LINKED ISSUE #4626: Implement Message Counter Synchronization Protocol (MCSP).
Problem
Peer message counter synchronization is an essential part of enabling secure CHIP messaging between members of an application group.

Proposed Solution
Implement  the protocol used to synchronize message counters between peers when using the same symmetric group key. This protocol is focused on encrypted messages sent over UDP with or without use of CHIP Reliable Messaging Protocol (CRMP).

This ticket does not handle the Initiating Message Counter Synchronization part which is depending on the following issues:
1. https://github.com/project-chip/connectedhomeip/issues/4628
2. https://github.com/project-chip/connectedhomeip/pull/4571

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
Peer message counter synchronization is an essential part of enabling secure CHIP messaging between members of an application group.

 #### Summary of Changes
Implement the protocol used to synchronize message counters between peers when using the same symmetric group key. 

This PR does not handle the Initiating Message Counter Synchronization part which is depending on the following issues:
#4628
#4571

 Fixes #4626

### CODE REVIEW

File: src/messaging/ReliableMessageMgr.cpp [hunk 1]
```diff
@@ -490,6 +491,29 @@ void ReliableMessageMgr::FailRetransTableEntries(ReliableMessageContext * rc, CH
...
 }
 
+/**
+ *  Retransmit all pending messages that were encrypted with application
+ *  group key and were addressed to the specified node.
+ *
+ *  @param[in] peerNodeId    Node ID of the destination node.
+ *
+ */
+void ReliableMessageMgr::RetransPendingAppGroupMsgs(uint64_t peerNodeId)
```
Comment (andy31415) [inline src/messaging/ReliableMessageMgr.cpp:501]: Please use full names like: `RetransmitPendingApplicationGroupMessages`

File: src/messaging/SecureChannelMgr.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
```
Comment (andy31415) [inline src/messaging/SecureChannelMgr.cpp:3]: Please remove copyright line (years are annoying to update).

@woody-apple  - is this ok with you? I find updating years tedious.

File: src/messaging/SecureChannelMgr.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements the CHIP Secure Channel protocol.
```
Comment (andy31415) [inline src/messaging/SecureChannelMgr.cpp:20]: Do we need this @file ?

File: src/messaging/tests/TestSecureChannelMgr.cpp [hunk 4]
```diff
@@ -0,0 +1,272 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
```
Comment (andy31415) [inline src/messaging/tests/TestSecureChannelMgr.cpp:3]: If keeping this should be just 2021 probably.

However I would still remove both Copyright and All rights reserved lines.

File: src/messaging/tests/TestSecureChannelMgr.cpp [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements unit tests for the ExchangeManager implementation.
```
Comment (andy31415) [inline src/messaging/tests/TestSecureChannelMgr.cpp:21]: I would drop this `@file` comment: generally I have not found any of these `@file` comments valuable when reading code and they seem to have a high probability of being copy & pasted - like this one I believe it should not be ExchangeManager but rather SecureChannelManager.

A comment saying that 'TestSecureChannelMgr.cpp implements unit tests for SecureChannelManager' seems redundant.

**Review by andy31415 — APPROVED** [review]

File: src/messaging/ReliableMessageMgr.cpp (hunk 1 above)
Comment (yufengwangca) [reply src/messaging/ReliableMessageMgr.cpp]: This function might be left to the later part of MCSP since it is not called in this PR

File: src/messaging/SecureChannelMgr.cpp (hunk 2 above)
Comment (yufengwangca) [reply src/messaging/SecureChannelMgr.cpp:3]: I prefer a separate PR to update the Copyright header part unanimously after we finalize what should be kept and what should be removed.

File: src/messaging/ExchangeContext.h [hunk 6]
```diff
@@ -179,6 +188,10 @@ class DLL_EXPORT ExchangeContext : public ReferenceCounted<ExchangeContext, Exch
     SecureSessionHandle mSecureSession; // The connection state
     uint16_t mExchangeId;               // Assigned exchange ID.
 
+    // [TODO: #4711]: this field need to be moved to appState object which implement 'exchange-specific' contextual
+    // actions with a delegate pattern.
+    uint8_t mChallenage[kMsgCounterChallengeSize]; // Challenge number to identify the sychronization request cryptographically.
```
Comment (mspang) [inline src/messaging/ExchangeContext.h:193]: typo: mChallenage -> mChallenge

File: src/messaging/SecureChannelMgr.cpp [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+#include <support/BufferWriter.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace SecureChannel {
+
+SecureChannelMgr::SecureChannelMgr()
+{
+    mExchangeMgr = nullptr;
```
Comment (mspang) [inline src/messaging/SecureChannelMgr.cpp:41]: can you init this in the class definition?

File: src/messaging/SecureChannelMgr.cpp [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+void SecureChannelMgr::Shutdown()
+{
+    if (mExchangeMgr != nullptr)
+    {
+        mExchangeMgr->UnregisterUnsolicitedMessageHandlerForProtocol(Protocols::kProtocol_SecureChannel);
+        mExchangeMgr = nullptr;
+    }
+}
+
+void SecureChannelMgr::OnMessageReceived(Messaging::ExchangeContext * ec, const PacketHeader & packetHeader,
```
Comment (mspang) [inline src/messaging/SecureChannelMgr.cpp:69]: Could we do ec -> exchangeContext ?

File: src/messaging/ExchangeContext.h (hunk 6 above)
Comment (yufengwangca) [reply src/messaging/ExchangeContext.h]: Thanks you!

File: src/messaging/SecureChannelMgr.cpp (hunk 8 above)
Comment (yufengwangca) [reply src/messaging/SecureChannelMgr.cpp]: acked

File: src/messaging/SecureChannelMgr.cpp (hunk 7 above)
Comment (yufengwangca) [reply src/messaging/SecureChannelMgr.cpp]: Acked

**Comment by woody-apple** [comment]
@pan-apple ?

**Comment by pan-apple** [comment]
@yufengwangca , the file name `SecureChannelMgr` seems to specifically implement `MessageCounterSync` protocol. The current file name is confusing, given we already have `SecureSessionMgr` class in the code base. How do you feel about renaming  `SecureChannelMgr.cpp/.h` to `MessageCounterSync.cpp/.h`?

File: src/messaging/ReliableMessageMgr.cpp [hunk 9]
```diff
@@ -25,6 +25,7 @@
 
 #include <messaging/ReliableMessageMgr.h>
 
+#include <core/CHIPKeyIds.h>
```
Comment (pan-apple) [inline src/messaging/ReliableMessageMgr.cpp:28]: Why is this change needed?

**Comment by yufengwangca** [comment]
> @yufengwangca , the file name `SecureChannelMgr` seems to specifically implement `MessageCounterSync` protocol. The current file name is confusing, given we already have `SecureSessionMgr` class in the code base. How do you feel about renaming `SecureChannelMgr.cpp/.h` to `MessageCounterSync.cpp/.h`?

Acked, SecureChannelMgr was originally designed to run MCSP and CASE protocols, since we have figured out how to implement multiple CASE sessions using ephemeral TCP ports within Transport layer, we can rename SecureChannelMgr to MessageCounterSyncMgr

File: src/messaging/ReliableMessageMgr.cpp (hunk 9 above)
Comment (yufengwangca) [reply src/messaging/ReliableMessageMgr.cpp]: Removed, this is for initiating MCSP, should be in the follow-up PR. Thanks!

**Comment by yufengwangca** [comment]
@pan-apple @woody-apple ?

File: src/messaging/ExchangeMgr.h [hunk 10]
```diff
@@ -177,6 +178,8 @@ class DLL_EXPORT ExchangeManager : public SecureSessionMgrDelegate
 
     ReliableMessageMgr * GetReliableMessageMgr() { return &mReliableMessageMgr; };
 
+    MessageCounterSyncMgr * GetSecureChannelMgr() { return &mSecureChannelMgr; };
```
Comment (pan-apple) [inline src/messaging/ExchangeMgr.h:181]: Could we also rename the function and the associated member variables?

File: src/messaging/MessageCounterSync.cpp [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+#include <support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace Messaging {
+
+CHIP_ERROR MessageCounterSyncMgr::Init(Messaging::ExchangeManager * exchangeMgr)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(exchangeMgr != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
```
Comment (pan-apple) [inline src/messaging/MessageCounterSync.cpp:43]: Some of the functions, where the `exit` block is just returning error, can be updated to use `ReturnMacros`. This will help reduce the boilerplate code as well.

`VerifyOrExit` can be replaced with `VerifyOrReturnError`
`SuccessOrExit` can be replaced with `ReturnErrorOnFailure`.

File: src/messaging/ExchangeContext.h [hunk 12]
```diff
@@ -33,6 +33,9 @@
 #include <transport/SecureSessionMgr.h>
 
 namespace chip {
+
+constexpr uint16_t kMsgCounterChallengeSize = 8; // The size of the message counter synchronization request message.
```
Comment (pan-apple) [inline src/messaging/ExchangeContext.h:37]: Should we move this to `MessageCounterSync.h`?

**Review by pan-apple — APPROVED** [review]
@yufengwangca , there are some minor comments. Once those are addressed, it looks good to me.

File: src/messaging/ExchangeMgr.h (hunk 10 above)
Comment (yufengwangca) [reply src/messaging/ExchangeMgr.h]: Acked

File: src/messaging/MessageCounterSync.cpp (hunk 11 above)
Comment (yufengwangca) [reply src/messaging/MessageCounterSync.cpp]: Acked

File: src/messaging/ExchangeContext.h (hunk 12 above)
Comment (yufengwangca) [reply src/messaging/ExchangeContext.h:37]: This will cause mutual including during compiling. 

The mChallenge should not be declared in ExchangeContext.h, it will be moved to MessageCounterSync.h after:
// [TODO: #4711]: this field need to be moved to appState object which implement 'exchange-specific' contextual actions with a delegate pattern.

**Review by jelderton — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]
